src/registry.rs

```diff
@@ -0,0 +1,164 @@
+                        new = n;
+                    }
+                }
+            }
+        })
+    }
+}
+
+thread_local! {
+    /// The thread registration harness.
+    ///
+    /// The harness is lazily initialized on its first use, thus registrating the current thread.
+    /// If initialized, the harness will get destructed on thread exit, which in turn unregisters
+    /// the thread.
+    static HARNESS: Harness = Harness {
+        entry: register(),
+    };
+}
+
+/// Holds a registered entry and unregisters it when dropped.
+struct Harness {
+    entry: *const Entry,
+}
+
+impl Drop for Harness {
+    fn drop(&mut self) {
+        unsafe {
+            let entry = &*self.entry;
+
+            // Unregister the thread by marking this entry as deleted.
+            unprotected(|scope| { entry.next.fetch_or(1, Release, scope); });
+        }
+    }
+}
+
+/// Returns a reference to the head pointer of the list of participating threads.
+fn participants() -> &'static Atomic<Entry> {
+    static PARTICIPANTS: AtomicUsize = ATOMIC_USIZE_INIT;
+    unsafe { &*(&PARTICIPANTS as *const AtomicUsize as *const Atomic<Entry>) }
+}
+
+/// Acquires a reference to the current participant.
+///
+/// Participants are lazily initialized on the first use.
+///
+/// # Panics
+///
+/// If this function is called while the thread is exiting, it might panic because it accesses
+/// thread-local data.
+pub fn with_current<F, R>(f: F) -> R
+where
+    F: FnOnce(&Participant) -> R,
+{
+    HARNESS.with(|harness| {
+        let entry = unsafe { &*harness.entry };
+        f(&entry.participant)
+    })
+}
+
+/// Returns an iterator over all participating threads.
+///
+/// Note that the iterator might return the same participant multiple times.
+pub fn iter(scope: &Scope) -> Iter {
+    let pred = participants();
+    let curr = pred.load(Acquire, scope);
+    Iter { scope, pred, curr }
+}
+
+/// An iterator over all participating threads.
+///
+/// Note that the iterator might return the same participant multiple times.
+pub struct Iter<'scope> {
+    /// The scope in which the iterator is operating.
+    scope: &'scope Scope,
+
+    /// Pointer from the predecessor to the current entry.
+    pred: &'scope Atomic<Entry>,
+
+    /// The current entry.
+    curr: Ptr<'scope, Entry>,
+}
+
+impl<'scope> Iterator for Iter<'scope> {
+    type Item = &'scope Participant;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(c) = unsafe { self.curr.as_ref() } {
+            let succ = c.next.load(Acquire, self.scope);
+
+            if succ.tag() == 1 {
+                // This thread has exited. Try unlinking it from the list.
+                let succ = succ.with_tag(0);
+
+                if self.pred
+                    .compare_and_set(self.curr, succ, AcqRel, self.scope)
+                    .is_err()
+                {
+                    // We lost the race to unlink this entry. There's no option other than to
```

Comment:
I think we could do simply reread it. Then, if the freshly read pointer is tagged with 1 as well, we must restart from the beginning. Not sure if that's worth the additional several lines of code, though...